tests/unit/s2n_self_talk_inet_socket_test.c

```diff
@@ -0,0 +1,202 @@
+    EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS12);
+
+    RESULT_GUARD(socket_cb->server_post_handshake_cb(server_conn));
+
+    char sync = 0;
+    char send_buffer[1] = { 0 };
+
+    send_buffer[0] = CHAR_A;
+    EXPECT_SUCCESS(s2n_send(server_conn, send_buffer, 1, &blocked));
+    EXPECT_SUCCESS(write(write_pipe, &sync, 1));
+
+    send_buffer[0] = CHAR_B;
+    EXPECT_SUCCESS(s2n_send(server_conn, send_buffer, 1, &blocked));
+    EXPECT_SUCCESS(write(write_pipe, &sync, 1));
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT launch_test(const struct self_talk_inet_socket_callbacks *socket_cb)
+{
+    /* configure real socket */
+    int listener = socket(AF_INET, SOCK_STREAM, 0);
+    RESULT_GUARD_POSIX(listener);
+    struct sockaddr_in saddr = { 0 };
+    saddr.sin_family = AF_INET;
+    saddr.sin_addr.s_addr = htonl(S2N_TEST_INADDR_LOOPBACK);
+    saddr.sin_port = 0;
+
+    /* listen on socket address */
+    socklen_t addrlen = sizeof(saddr);
+    RESULT_GUARD_POSIX(bind(listener, (struct sockaddr *) &saddr, addrlen));
+    RESULT_GUARD_POSIX(getsockname(listener, (struct sockaddr *) &saddr, &addrlen));
+
+    /* Used for synchronizing read and writes between client and server.
+     *
+     * The purpose of this mechanism is to prevent the client from reading
+     * before the server has sent it. The client blocks on reading a single
+     * char while the server writes a char after it has sent data and wishes
+     * to unblock the client. */
+    int sync_pipe[2] = { 0 };
+    RESULT_GUARD_POSIX(pipe(sync_pipe));
+
+    pid_t child = fork();
+    RESULT_ENSURE(child >= 0, S2N_ERR_SAFETY);
+    int status = 0;
+    int fd = 0;
+    if (child) {
+        /* server */
+        RESULT_GUARD_POSIX(listen(listener, 1));
+        fd = accept(listener, NULL, NULL);
+        RESULT_GUARD_POSIX(fd);
+
+        RESULT_GUARD_POSIX(close(sync_pipe[0]));
+        RESULT_GUARD(start_server(fd, sync_pipe[1], socket_cb));
+
+        RESULT_ENSURE_EQ(waitpid(-1, &status, 0), child);
+        RESULT_ENSURE_EQ(status, 0);
+    } else {
+        /* client */
+        fd = socket(AF_INET, SOCK_STREAM, 0);
+        RESULT_GUARD_POSIX(fd);
+
+        /* wait for server to start up */
+        sleep(1);
+        RESULT_GUARD_POSIX(connect(fd, (struct sockaddr *) &saddr, addrlen));
+
+        RESULT_GUARD_POSIX(close(sync_pipe[1]));
+        RESULT_GUARD(start_client(fd, sync_pipe[0], socket_cb));
+        exit(0);
+    }
+
+    return S2N_RESULT_OK;
+}
+
+/* These test are unique compared to our other self talk tests because they are testing
+ * socket functionality. When enabled, kTLS offloads the TLS protocol to the socket so
+ * its necessary to use AF_INET sockets to test this functionality. Its not possible to
+ * enable kTLS on a AF_UNIX socket, which is what our other self talk test use. The use
+ * of AF_INET socket also drives the need to fork and create two process (client and
+ * server) to establish the connection.
+ */
```

Comment:
```suggestion
/* These tests are unique compared to our other self talk tests because they are testing
 * socket functionality. When enabled, kTLS offloads the TLS protocol to the socket so
 * it's necessary to use AF_INET sockets to test this functionality. It's not possible to
 * enable kTLS on a AF_UNIX socket, which is what our other self talk tests use. The use
 * of an AF_INET socket also drives the need to fork and create two process (client and
 * server) to establish the connection.
 */
```